internal/terraform/context_apply2_test.go

```diff
@@ -2456,3 +2455,93 @@ resource "test_object" "foo" {
 		t.Errorf("expected:\n%s\nactual:\n%s\ndiff:\n%s", expected, string(instance.Current.AttrsJSON), diff)
 	}
 }
+
+func TestContext2Apply_forget(t *testing.T) {
```

Comment:
The presence of this test causes the fails in `TestContext2Apply_errorRestore*`. I'm not sure exactly what's going wrong - running `ctx.Plan()` with the mock provider seems to be sufficient.

Reply:
I'm not sure what's going on here either, but FWIW I compared the logs of `TestContext2Apply_errorRestoreStatus` with and without this "forget" test running and see that it seems to end up planning "forget" in a bunch of places where it would previously plan "delete".

Here's a subset:

```diff
- [TRACE] writeChange: recorded Delete change for test_object.a
+ [TRACE] writeChange: recorded Forget change for test_object.a
```

```diff
- [TRACE] DiffTransformer: found Delete change for test_object.a
- [TRACE] DiffTransformer: test_object.a will be represented for destruction by test_object.a (destroy)
+ [TRACE] DiffTransformer: found Forget change for test_object.a
+ [TRACE] DiffTransformer: test_object.a will be represented for forgetting by test_object.a (forget)
```

```diff
- [DEBUG] test_object.a: applying the planned Delete change
- [TRACE] vertex "test_object.a (destroy)": visit complete, with errors
+ [TRACE] vertex "test_object.a (forget)": visit complete
```

I notice that both of these tests include a `test_object.a`, so it seems like somehow the "removed statement" is persisting from one `terraform.Context` into the next, making all subsequent tests treat the removal of `test_object.a` as a `destroy = false`. :thinking:

Reply:
I notice that the call below is passing `DefaultPlanOpts` in to `ctx.Plan`, and the other test does too. That variable is a pointer to a shared object, so if `Context.Plan` were to modify the given plan opts then it would modify the shared `DefaultPlanOpts` object, thereby changing what's used by default in later tests using that object. The other test uses it too.

Therefore my guess here is that you've got some logic that's directly modifying the plan opts based on what's found in the configuration, and that's therefore changing what the default is.

Reply:
Thanks! It looks like my laziness in naming resources has exposed an existing bug in context_plan.go, where `opts` is modified repeatedly. Fixing...

Reply:
Implementing a `DeepCopy()` for `PlanOpts` is a large undertaking because of the number of nested types in there that will need their own deep-copy behaviour. I'm going to follow up with this in a separate PR.

For the moment I'll modify the tests to sidestep the shared PlanOpts issue. Since `ctx.Plan` can only be called with a shared PlanOpts object in test code, I think this is acceptable.

Reply:
I think https://github.com/hashicorp/terraform/pull/34355 is a more sensible fix - remove these fields from `PlanOpts` so there is no need to modify it inside `ctx.plan`.